kubernetes/provider.go

```diff
@@ -234,7 +233,47 @@ func (k kubeClientsets) AggregatorClientset() (*aggregator.Clientset, error) {
 	return k.aggregatorClientset, nil
 }
 
+func inCluster() bool {
+	host, port := os.Getenv("KUBERNETES_SERVICE_HOST"), os.Getenv("KUBERNETES_SERVICE_PORT")
+	return host != "" && port != ""
+}
+
+var authDocumentationURL = "https://registry.terraform.io/providers/hashicorp/kubernetes/latest/docs#authentication"
+
+func checkConfigurationValid(d *schema.ResourceData) error {
+	if inCluster() {
+		log.Printf("[DEBUG] Terraform appears to be running inside the Kubernetes cluster")
+		return nil
+	}
+
+	if os.Getenv("KUBE_CONFIG_PATHS") != "" {
+		return nil
+	}
+
+	atLeastOneOf := []string{
+		"host",
+		"config_path",
+		"config_paths",
+		"client_certificate",
+		"token",
+		"exec",
+	}
+	for _, a := range atLeastOneOf {
```

Comment:
I'm not sure I understand this correctly, but it looks like having just "token" or "client_certificate" set would count as valid configuration. Not sure if I'm missing something else.

This could be confusing if not covering all combinations that make a valid configuration.

Also (not 100% sure) but it might be falsely triggered by progressive apply situations. Can we rule those out?

Reply:
The intent of this code is not to cover all valid possible configurations, it is to rule out that:

1. The user has left the provider block is empty 
2. Terraform is running inside a cluster. 
 
The purpose of this is so we can tell the user to explicitly configure instead of generating the confusing "could not connect" error at plan time when the client actually gets initialized. Once we fix [this](https://github.com/hashicorp/terraform-provider-kubernetes/blob/master/kubernetes/provider.go#L245) we should be able to generate more meaningful errors that come from client go, as it will warn you about these problems.

We can use `ConflictsWith` and `RequiredWith` to validate that the right combination of attributes is being used together, but I would propose doing that in another pull request.

Reply:
Sounds good. I didn't catch those nuances while reading through it the first time. Let try this approach out and see how it improves things. Looks like https://github.com/hashicorp/terraform/issues/24055 is confirmed fixed so we're clear to move in that regard.
I agree with doing things incrementally. Let's re-evaluate after this is released.